### PR TITLE
initial_unit_tests

### LINKED ISSUE #83: implement unit tests and run all tests target to test for regressions
Pytest unit tests can easily be written for a large amount of the codebase. A run all tests target, likely in the pyproject, should run everything in the tests folder to quickly test changes for regressions.

### PR DESCRIPTION
**Pull request by arades79**
### Summary

This PR begins work on unit testing all of hyperdome. This lays a base for future changes to implement unit tests on code they change. Unit tests for most of the common modules were written, and can serve as examples of very robust tests for future unit tests to go by.

Much of the testing is done with the hypothesis library, which essentially acts as a fuzzer and effectively multiplying the number of tests executed. Future tests should also strive to leverage this library.

Coverage was also added, and a rudementary coverage config was added to the pyproject.toml. Current coverage is 9%, with 3 of the 4 test files written having 100% coverage of their given modules.

Tests are best run using `poetry run coverage run` in the project root, this way you can get a coverage report with `poetry run coverage report`. Otherwise tests can be run without coverage recording using `poetry run pytest .`.

This PR fixes #83 
### Checklist

- [x] If code changes were made then they have been tested.
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: hyperdome/common/common.py [hunk 1]
```diff
@@ -107,20 +107,36 @@ def tor_paths() -> typing.Tuple[Path, Path, Path, Path]:
...
 def get_available_port(min_port: int, max_port: int) -> int:
     """
     Find a random available port within the given range.
     """
+    if not (isinstance(min_port, int) and isinstance(max_port, int)):
+        raise TypeError("ports must be integers")
+    if not (
+        MIN_PORT < min_port < MAX_PORT
+        and MIN_PORT < max_port < MAX_PORT
+        and min_port < max_port - 1
```
Comment (stevenpitts) [inline hyperdome/common/common.py]: Couldn't this be `MIN_PORT < min_port <= max_port < MAX_PORT`?

File: poetry.lock [hunk 2]
```diff
@@ -124,6 +124,22 @@ optional = false
 python-versions = ">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*, !=3.4.*"
 version = "0.4.3"
 
+[[package]]
+category = "dev"
+description = "Code coverage measurement for Python"
+name = "coverage"
+optional = false
+python-versions = ">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*, !=3.4.*, <4"
```
Comment (stevenpitts) [inline poetry.lock:132]: Futureproofing I see, I like it

I can't wait until we get to deprecate Python3 :D

File: test/hyperdome_common_test/test_available_port.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+MIN_USED_PORT = 62105
+
+MAX_UNUSED_PORT = 63125
+MIN_UNUSED_PORT = 63105
+
+
+@pytest.fixture(scope="session")
+def bound_ports():
+    sockets = []
+    for i in range(MIN_USED_PORT, MAX_USED_PORT):
```
Comment (stevenpitts) [inline test/hyperdome_common_test/test_available_port.py]: Should this be `(MIN_USED_PORT, MAX_USED_PORT+1)`? Not that it really matters

File: test/hyperdome_common_test/test_available_port.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    with pytest.raises(TypeError) as e:
+        get_available_port(min_port, 65000)
+
+    with pytest.raises(TypeError) as e:
+        get_available_port(1000, max_port)
+
+
+@given(min_port=st.just(MIN_USED_PORT), max_port=st.just(MAX_USED_PORT))
+def test_bound_ports_error(bound_ports, min_port, max_port):
+    bound_ports
```
Comment (stevenpitts) [inline test/hyperdome_common_test/test_available_port.py]: :eyes:

File: test/hyperdome_common_test/test_bootstrap.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import hyperdome.common.bootstrap as bs
```
Comment (stevenpitts) [inline test/hyperdome_common_test/test_bootstrap.py:22]: this is bs

File: test/hyperdome_common_test/test_encryption.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+def test_rotation_cannot_decrypt(
+    pre_exchanged_user_factory: user_fixture, message: str
+):
+    user_1_crypto, user_2_crypto = pre_exchanged_user_factory()
+
+    enc_message = user_1_crypto.encrypt_outgoing_message(message)
+
+    prev_chat_key = user_2_crypto._chat_key
+
+    _ = user_2_crypto.public_chat_key
```
Comment (stevenpitts) [inline test/hyperdome_common_test/test_encryption.py:103]: Is this to cycle the public chat key? I'm not entirely sure I get it

**Review by stevenpitts — DISMISSED** [review]
Super glad unit tests are here! Even if not totally done, it's got a lot of coverage

File: hyperdome/common/common.py (hunk 1 above)
Comment (arades79) [reply hyperdome/common/common.py]: I figured there was a better way, let me check if that still passes all tests.

File: test/hyperdome_common_test/test_available_port.py (hunk 3 above)
Comment (arades79) [reply test/hyperdome_common_test/test_available_port.py]: yeah it, it needs to be inclusive of MAX, it's probably clearer if it's in the for.

File: test/hyperdome_common_test/test_available_port.py (hunk 4 above)
Comment (arades79) [reply test/hyperdome_common_test/test_available_port.py]: what, I didn't want an unused argument warning 😋

File: test/hyperdome_common_test/test_encryption.py (hunk 6 above)
Comment (arades79) [reply test/hyperdome_common_test/test_encryption.py:103]: yes, public_chat_key is a property, and values provided by it should only ever be used a single time, so it's called here to force a key rotation.

File: test/hyperdome_common_test/test_available_port.py (hunk 4 above)
Comment (stevenpitts) [reply test/hyperdome_common_test/test_available_port.py]: Ah, probably worth a comment tho

**Review by stevenpitts — APPROVED** [review]

**Pull request merged**